# Introduction to Natural Language Processing (NLP)

Generally speaking, <i>Computational Text Analysis</i> is a set of interpretive methods which seek to understand patterns in human discourse, in part through statistics. More familiar methods, such as close reading, are exceptionally well-suited to the analysis of individual texts, however our research questions typically compel us to look for relationships across texts, sometimes counting in the thousands or even millions. We have to zoom out, in order to perform so-called <i>distant reading</i>. Fortunately for us, computers are well-suited to identify the kinds of textual relationships that exist at scale.

We will spend the week exploring research questions that computational methods can help to answer and thinking about how these complement -- rather than displace -- other interpretive methods. Before moving to that conceptual level, however, we will familiarize ourselves with the basic tools of the trade.

<i>Natural Language Processing</i> is an umbrella term for the methods by which a computer handles human language text. This includes transforming the text into a numerical form that the computer manipulates natively, as well as the measurements that reserchers often perform. In the parlance, <i>natural language</i> refers to a language spoken by humans, as opposed to a <i>formal language</i>, such as Python, which comprises a set of logical operations.

The goal of this lesson is to jump right in to text analysis and natural language processing. Rather than starting with the nitty gritty of programming in Python, this lesson will demonstrate some neat things you can do with a minimal amount of coding. Today, we aim to build intuition about how computers read human text and learn some of the basic operations we'll perform with them.


# Lesson Outline
- Jargon
- Text in Python
- Tokenization & Term Frequency
- Pre-Processing: 
    * Changing words to lowercase
    * Removing stop words
    * Removing punctuation
- Part-of-Speech Tagging
    * Tagging tokens
    * Counting tagged tokens
- Demonstration: Guess the Novel!
- Concordance

# 0. Key Jargon

### General
* *programming* (or *coding*)
    * A program is a sequence of instructions given to the computer, in order to perform a specific task. Those instructions are written in a specific *programming language*, in our case, Python. Writing these instructions can be an art as much as a science.
* *Python*
    * A general-use programming language that is popular for NLP and statistics.
* *script*
    * A block of executable code.
* *Jupyter Notebook*
    * Jupyter is a popular interface in which Python scripts can be written and executed. Stand-alone scripts are saved in Notebooks. The script can be sub-divided into units called <i>cells</i> and executed individually. Cells can also contain discursive text and html formatting (such as in this cell!)
* *package* (or *module*)
    * Python offers a basic set of functions that can be used off-the-shelf. However, we often wish to go beyond the basics. To that end, <i>packages</i> are collections of python files that contain pre-made functions. These functions are made available to our program when we <i>import</i> the package that contains them.
* *Anaconda*
    * Anaconda is a <i>platform</i> for programming in Python. A platform constitutes a closed environment on your computer that has been standardized for functionality. For example, Anaconda contains common packages and programming interfaces for Python, and its developers ensure compatibility among the moving parts.
    
    
### When Programming
* *variable*
    * A variable is a generic container that stores a value, such as a number or series of letters. This is not like a variable from high-school algebra, which had a single "correct" value that must be solved. Rather, the user <i>assigns</i> values to the variable in order to perform operations on it later.
* *string*
    * A type of object consisting of a single sequence of alpha-numeric characters. In Python, a string is indicated by quotation marks around the sequence"
* *list*
    * A type of object that consists of a sequence of elements.
    
### Natural Language Processing
* *pre-processing*
    * Transforming a human lanugage text into computer-manipulable format. A typical pre-processing workflow includes <i>stop-word</i> removal, setting text in lower case, and <i>term frequency</i> counting.
* *token*
    * An individual word unit within a sentence.
* *stop words*
    * The function words in a natural langauge, such as <i>the</i>, <i>of</i>, <i>it</i>, etc. These are typically the most common words.
* *term frequency*
    * The number of times a term appears in a given text. This is either reported as a raw tally or it is <i>normalized</i> by dividing by the total number of words in a text.    
* *POS tagging*
    * One common task in NLP is the determination of a word's part-of-speech (POS). The label that describes a word's POS is called its <i>tag</i>. Specialized functions that make these determinations are called <i>POS Taggers</i>.
* *concordance*
    * Index of instances of a given word (or other linguistic feature) in a text. Typically, each instance is presented within a contextual window for human readability.
* *NLTK (Natural Language Tool Kit)*
    * A common Python package that contains many NLP-related functions

### Further Resources:

Check out the full range of techniques included in Python's nltk package here: http://www.nltk.org/book/

# 1. Text in Python

First, a quote about what digital humanities means, from digital humanist Kathleen Fitzpatrick. Source: "On Scholarly Communication and the Digital Humanities: An Interview with Kathleen Fitzpatrick", *In the Library with the Lead Pipe*

In [1]:
print("For me it has to do with the work that gets done at the crossroads of digital media and traditional humanistic study. And that happens in two different ways. On the one hand, it's bringing the tools and techniques of digital media to bear on traditional humanistic questions; on the other, it's also bringing humanistic modes of inquiry to bear on digital media.")

For me it has to do with the work that gets done at the crossroads of digital media and traditional humanistic study. And that happens in two different ways. On the one hand, it's bringing the tools and techniques of digital media to bear on traditional humanistic questions; on the other, it's also bringing humanistic modes of inquiry to bear on digital media.


In [2]:
# Assign the quote to a variable, so we can refer back to it later
# We get to make up the name of our variable, so let's give it a descriptive label: "sentence"

sentence = "For me it has to do with the work that gets done at the crossroads of digital media and traditional humanistic study. And that happens in two different ways. On the one hand, it's bringing the tools and techniques of digital media to bear on traditional humanistic questions; on the other, it's also bringing humanistic modes of inquiry to bear on digital media."

In [3]:
# Oh, also: anything on a line starting with a hashtag is called a comment,
# and is meant to clarify code for human readers. The computer ignores these lines.

In [4]:
# Print the contents of the variable 'sentence'

print(sentence)

For me it has to do with the work that gets done at the crossroads of digital media and traditional humanistic study. And that happens in two different ways. On the one hand, it's bringing the tools and techniques of digital media to bear on traditional humanistic questions; on the other, it's also bringing humanistic modes of inquiry to bear on digital media.


# 2. Tokenizing Text and Counting Words

The above output is how a human would read that sentence. Next we look the main way in which a computer "reads", or *parses*, that sentence.

The first step is typically to <i>tokenize</i> it, or to change it into a series of <i>tokens</i>. Each token roughly corresponds to either a word or punctuation mark. These smaller units are more straight-forward for the computer to handle for tasks like counting.

In [5]:
# Import the NLTK (Natural Language Tool Kit) package

import nltk

In [6]:
# Tokenize our sentence!

nltk.word_tokenize(sentence)

['For',
 'me',
 'it',
 'has',
 'to',
 'do',
 'with',
 'the',
 'work',
 'that',
 'gets',
 'done',
 'at',
 'the',
 'crossroads',
 'of',
 'digital',
 'media',
 'and',
 'traditional',
 'humanistic',
 'study',
 '.',
 'And',
 'that',
 'happens',
 'in',
 'two',
 'different',
 'ways',
 '.',
 'On',
 'the',
 'one',
 'hand',
 ',',
 'it',
 "'s",
 'bringing',
 'the',
 'tools',
 'and',
 'techniques',
 'of',
 'digital',
 'media',
 'to',
 'bear',
 'on',
 'traditional',
 'humanistic',
 'questions',
 ';',
 'on',
 'the',
 'other',
 ',',
 'it',
 "'s",
 'also',
 'bringing',
 'humanistic',
 'modes',
 'of',
 'inquiry',
 'to',
 'bear',
 'on',
 'digital',
 'media',
 '.']

In [7]:
# Create new variable that contains our tokenized sentence

sentence_tokens = nltk.word_tokenize(sentence)

In [8]:
# Inspect our new variable
# Note the square braces at the beginning and end that indicate we are looking at a list-type object

print(sentence_tokens)

['For', 'me', 'it', 'has', 'to', 'do', 'with', 'the', 'work', 'that', 'gets', 'done', 'at', 'the', 'crossroads', 'of', 'digital', 'media', 'and', 'traditional', 'humanistic', 'study', '.', 'And', 'that', 'happens', 'in', 'two', 'different', 'ways', '.', 'On', 'the', 'one', 'hand', ',', 'it', "'s", 'bringing', 'the', 'tools', 'and', 'techniques', 'of', 'digital', 'media', 'to', 'bear', 'on', 'traditional', 'humanistic', 'questions', ';', 'on', 'the', 'other', ',', 'it', "'s", 'also', 'bringing', 'humanistic', 'modes', 'of', 'inquiry', 'to', 'bear', 'on', 'digital', 'media', '.']


### Note on Tokenization

While seemingly simple, tokenization is a non-trivial task.

For example, notice how the tokenizer has handled contractions: a contracted word is divided into two separate tokens! What do you think is the motivation for this? How else might you tokenize them?

Also notice each token is either a word or punctuation mark. In practice, it is sometimes useful to remove punctuation marks and at other times to include them, depending on the situation.

In the coming days, we will see other tokenizers and have opportunities to explore their reasoning. For now, we will look at a few examples of NLP tasks that tokenization enables.

In [9]:
# How many tokens are in our list?

len(sentence_tokens)

71

In [11]:
# How often does each token appear in our list?

import collections

collections.Counter(sentence_tokens)

Counter({"'s": 2,
         ',': 2,
         '.': 3,
         ';': 1,
         'And': 1,
         'For': 1,
         'On': 1,
         'also': 1,
         'and': 2,
         'at': 1,
         'bear': 2,
         'bringing': 2,
         'crossroads': 1,
         'different': 1,
         'digital': 3,
         'do': 1,
         'done': 1,
         'gets': 1,
         'hand': 1,
         'happens': 1,
         'has': 1,
         'humanistic': 3,
         'in': 1,
         'inquiry': 1,
         'it': 3,
         'me': 1,
         'media': 3,
         'modes': 1,
         'of': 3,
         'on': 3,
         'one': 1,
         'other': 1,
         'questions': 1,
         'study': 1,
         'techniques': 1,
         'that': 2,
         'the': 5,
         'to': 3,
         'tools': 1,
         'traditional': 2,
         'two': 1,
         'ways': 1,
         'with': 1,
         'work': 1})

In [12]:
# Assign those token counts to a variable

token_frequency = collections.Counter(sentence_tokens)

In [13]:
# Get an ordered list of the most frequent tokens

token_frequency.most_common(10)

[('the', 5),
 ('it', 3),
 ('to', 3),
 ('of', 3),
 ('digital', 3),
 ('media', 3),
 ('humanistic', 3),
 ('.', 3),
 ('on', 3),
 ('that', 2)]

### Note on Term Frequency

Some of the most frequent words appear to summarize the sentence: in particular the words "humanistic", "digital", and "media". However, most of the these terms seem to add noise in the summary: "the", "it", "to", ".", etc.

There are many strategies for identifying the most important words in a text, and we will cover the most popular ones in the next week. Today, we will look at two of them. In the first, we will simply remove the noisey tokens. In the second, we will identify important words using their parts of speech.

# 3. Pre-Processing: Lower Case, Remove Stop Words and Punctuation

Typically, a text goes through a number of pre-processing steps before beginning to the actual analysis. We have already seen the tokenization step. Typically, pre-processing includes transforming tokens to lower case and removing stop words and punctuation marks.

Again, pre-processing is a non-trivial process that can have large impacts on the analysis that follows. For instance, what will be the most common token in our example sentence, once we set all tokens to lower case?

### Lower Case

In [14]:
# Let's revisit our original sentence

sentence

"For me it has to do with the work that gets done at the crossroads of digital media and traditional humanistic study. And that happens in two different ways. On the one hand, it's bringing the tools and techniques of digital media to bear on traditional humanistic questions; on the other, it's also bringing humanistic modes of inquiry to bear on digital media."

In [15]:
# And now transform it to lower case, all at once

sentence.lower()

"for me it has to do with the work that gets done at the crossroads of digital media and traditional humanistic study. and that happens in two different ways. on the one hand, it's bringing the tools and techniques of digital media to bear on traditional humanistic questions; on the other, it's also bringing humanistic modes of inquiry to bear on digital media."

In [16]:
# Okay, let's set our list of tokens to lower case, one at a time

# The syntax of the line below is tricky. Don't worry about it for now.
# We'll spend plenty of time on it tomorrow!

lower_case_tokens = [ word.lower()  for word in sentence_tokens ]

In [17]:
# Inspect

print(lower_case_tokens)

['for', 'me', 'it', 'has', 'to', 'do', 'with', 'the', 'work', 'that', 'gets', 'done', 'at', 'the', 'crossroads', 'of', 'digital', 'media', 'and', 'traditional', 'humanistic', 'study', '.', 'and', 'that', 'happens', 'in', 'two', 'different', 'ways', '.', 'on', 'the', 'one', 'hand', ',', 'it', "'s", 'bringing', 'the', 'tools', 'and', 'techniques', 'of', 'digital', 'media', 'to', 'bear', 'on', 'traditional', 'humanistic', 'questions', ';', 'on', 'the', 'other', ',', 'it', "'s", 'also', 'bringing', 'humanistic', 'modes', 'of', 'inquiry', 'to', 'bear', 'on', 'digital', 'media', '.']


### Stop Words

In [18]:
# Import the stopwords list
from nltk.corpus import stopwords

In [19]:
# Take a look at what stop words are included

print(stopwords.words('english'))

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', 'her', 'hers', 'herself', 'it', 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', 'too', 'very', 's', 't', 'can', 'will', 'just', 'don', 'should', 'no

In [20]:
# Try another language

print(stopwords.words('spanish'))

['de', 'la', 'que', 'el', 'en', 'y', 'a', 'los', 'del', 'se', 'las', 'por', 'un', 'para', 'con', 'no', 'una', 'su', 'al', 'lo', 'como', 'más', 'pero', 'sus', 'le', 'ya', 'o', 'este', 'sí', 'porque', 'esta', 'entre', 'cuando', 'muy', 'sin', 'sobre', 'también', 'me', 'hasta', 'hay', 'donde', 'quien', 'desde', 'todo', 'nos', 'durante', 'todos', 'uno', 'les', 'ni', 'contra', 'otros', 'ese', 'eso', 'ante', 'ellos', 'e', 'esto', 'mí', 'antes', 'algunos', 'qué', 'unos', 'yo', 'otro', 'otras', 'otra', 'él', 'tanto', 'esa', 'estos', 'mucho', 'quienes', 'nada', 'muchos', 'cual', 'poco', 'ella', 'estar', 'estas', 'algunas', 'algo', 'nosotros', 'mi', 'mis', 'tú', 'te', 'ti', 'tu', 'tus', 'ellas', 'nosotras', 'vosostros', 'vosostras', 'os', 'mío', 'mía', 'míos', 'mías', 'tuyo', 'tuya', 'tuyos', 'tuyas', 'suyo', 'suya', 'suyos', 'suyas', 'nuestro', 'nuestra', 'nuestros', 'nuestras', 'vuestro', 'vuestra', 'vuestros', 'vuestras', 'esos', 'esas', 'estoy', 'estás', 'está', 'estamos', 'estáis', 'están', 

In [21]:
# Create a new variable that contains the sentence tokens but NOT the stopwords

tokens_nostops = [ word  for word in lower_case_tokens  if word not in stopwords.words('english') ]

In [22]:
# Inspect

print(tokens_nostops)

['work', 'gets', 'done', 'crossroads', 'digital', 'media', 'traditional', 'humanistic', 'study', '.', 'happens', 'two', 'different', 'ways', '.', 'one', 'hand', ',', "'s", 'bringing', 'tools', 'techniques', 'digital', 'media', 'bear', 'traditional', 'humanistic', 'questions', ';', ',', "'s", 'also', 'bringing', 'humanistic', 'modes', 'inquiry', 'bear', 'digital', 'media', '.']


### Punctuation

In [23]:
# Import a list of punctuation marks

import string

In [24]:
# Inspect

string.punctuation

'!"#$%&\'()*+,-./:;<=>?@[\\]^_`{|}~'

In [25]:
# Remove punctuation marks from token list

tokens_clean = [word for word in tokens_nostops if word not in string.punctuation]

In [26]:
# See what's left

print(tokens_clean)

['work', 'gets', 'done', 'crossroads', 'digital', 'media', 'traditional', 'humanistic', 'study', 'happens', 'two', 'different', 'ways', 'one', 'hand', "'s", 'bringing', 'tools', 'techniques', 'digital', 'media', 'bear', 'traditional', 'humanistic', 'questions', "'s", 'also', 'bringing', 'humanistic', 'modes', 'inquiry', 'bear', 'digital', 'media']


### Re-count the Most Frequent Words

In [27]:
# Count the new token list

word_frequency_clean = collections.Counter(tokens_clean)

In [28]:
# Most common words

word_frequency_clean.most_common(10)

[('digital', 3),
 ('media', 3),
 ('humanistic', 3),
 ('traditional', 2),
 ("'s", 2),
 ('bringing', 2),
 ('bear', 2),
 ('work', 1),
 ('gets', 1),
 ('done', 1)]

Better! The ten most frequent words now give us a pretty good sense of the substance of this sentence. But we still have problems. For example, the token "'s" sneaked in there. One solution is to keep adding stop words to our list, but this could go on forever and is not a good solution when processing lots of text.

There's another way of identifying content words, and it involves identifying the part of speech of each word.

# 4. Part-of-Speech Tagging

You may have noticed that stop words are typically short function words, like conjunctions and prepositions. Intuitively, if we could identify the part of speech of a word, we would have another way of identifying which contribute to the text's subject matter. NLTK can do that too!

NLTK has a <i>POS Tagger</i>, which identifies and labels the part-of-speech (POS) for every token in a text. The particular labels that NLTK uses come from the Penn Treebank corpus, a major resource from corpus linguistics.

You can find a list of all Penn POS tags here: https://www.ling.upenn.edu/courses/Fall_2003/ling001/penn_treebank_pos.html

Note that, from this point on, the code is going to get a little more complex. Don't worry about the particularities of each line. For now, we will focus on the NLP tasks themselves and the textual patterns they identify.

In [31]:
# Let's revisit our original list of tokens

print(sentence_tokens)

['For', 'me', 'it', 'has', 'to', 'do', 'with', 'the', 'work', 'that', 'gets', 'done', 'at', 'the', 'crossroads', 'of', 'digital', 'media', 'and', 'traditional', 'humanistic', 'study', '.', 'And', 'that', 'happens', 'in', 'two', 'different', 'ways', '.', 'On', 'the', 'one', 'hand', ',', 'it', "'s", 'bringing', 'the', 'tools', 'and', 'techniques', 'of', 'digital', 'media', 'to', 'bear', 'on', 'traditional', 'humanistic', 'questions', ';', 'on', 'the', 'other', ',', 'it', "'s", 'also', 'bringing', 'humanistic', 'modes', 'of', 'inquiry', 'to', 'bear', 'on', 'digital', 'media', '.']


In [34]:
# Use the NLTK POS tagger

nltk.pos_tag(sentence_tokens)

[('For', 'IN'),
 ('me', 'PRP'),
 ('it', 'PRP'),
 ('has', 'VBZ'),
 ('to', 'TO'),
 ('do', 'VB'),
 ('with', 'IN'),
 ('the', 'DT'),
 ('work', 'NN'),
 ('that', 'WDT'),
 ('gets', 'VBZ'),
 ('done', 'VBN'),
 ('at', 'IN'),
 ('the', 'DT'),
 ('crossroads', 'NNS'),
 ('of', 'IN'),
 ('digital', 'JJ'),
 ('media', 'NNS'),
 ('and', 'CC'),
 ('traditional', 'JJ'),
 ('humanistic', 'JJ'),
 ('study', 'NN'),
 ('.', '.'),
 ('And', 'CC'),
 ('that', 'DT'),
 ('happens', 'VBZ'),
 ('in', 'IN'),
 ('two', 'CD'),
 ('different', 'JJ'),
 ('ways', 'NNS'),
 ('.', '.'),
 ('On', 'IN'),
 ('the', 'DT'),
 ('one', 'CD'),
 ('hand', 'NN'),
 (',', ','),
 ('it', 'PRP'),
 ("'s", 'VBZ'),
 ('bringing', 'VBG'),
 ('the', 'DT'),
 ('tools', 'NNS'),
 ('and', 'CC'),
 ('techniques', 'NNS'),
 ('of', 'IN'),
 ('digital', 'JJ'),
 ('media', 'NNS'),
 ('to', 'TO'),
 ('bear', 'VB'),
 ('on', 'IN'),
 ('traditional', 'JJ'),
 ('humanistic', 'JJ'),
 ('questions', 'NNS'),
 (';', ':'),
 ('on', 'IN'),
 ('the', 'DT'),
 ('other', 'JJ'),
 (',', ','),
 ('it', 

In [35]:
# Assign POS-tagged list to a variable

tagged_tokens = nltk.pos_tag(sentence_tokens)

### Most Frequent POS Tags

In [39]:
# We'll tread lightly here, and just say that we're counting POS tags

tag_frequency = collections.Counter( [ tag for (word, tag) in tagged_tokens ])

In [40]:
# POS Tags sorted by frequency

tag_frequency.most_common()

[('IN', 11),
 ('JJ', 10),
 ('NNS', 9),
 ('DT', 6),
 ('VBZ', 5),
 ('PRP', 4),
 ('NN', 4),
 ('TO', 3),
 ('VB', 3),
 ('CC', 3),
 ('.', 3),
 ('CD', 2),
 (',', 2),
 ('VBG', 2),
 ('WDT', 1),
 ('VBN', 1),
 (':', 1),
 ('RB', 1)]

### Now it's getting interesting

The "IN" tag refers to prepositions, so it's no surprise that it should be the most common. However, we can see at a glance now that the sentence contains a lot of adjectives, "JJ". This feels like it tells us something about the rhetorical style or structure of the sentence: certain qualifiers seem to be important to the meaning of the sentence.

Let's dig in to see what those adjectives are.

In [42]:
# Let's filter our list, so it only keeps adjectives

adjectives = [word for word,pos in tagged_tokens if pos == 'JJ' or pos=='JJR' or pos=='JJS']

In [44]:
# Inspect

print( adjectives )

['digital', 'traditional', 'humanistic', 'different', 'digital', 'traditional', 'humanistic', 'other', 'humanistic', 'digital']


In [46]:
# Tally the frequency of each adjective

adj_frequency = collections.Counter(adjectives)

In [49]:
# Most frequent adjectives

adj_frequency.most_common(5)

[('digital', 3),
 ('humanistic', 3),
 ('traditional', 2),
 ('different', 1),
 ('other', 1)]

In [51]:
# Let's do the same for nouns.

nouns = [word for word,pos in tagged_tokens if pos=='NN' or pos=='NNS']

In [52]:
# Inspect

print(nouns)

['work', 'crossroads', 'media', 'study', 'ways', 'hand', 'tools', 'techniques', 'media', 'questions', 'modes', 'inquiry', 'media']


In [53]:
# Tally the frequency of the nouns

noun_frequency = collections.Counter(nouns)

In [54]:
# Most Frequent Nouns

print(noun_frequency.most_common(5))

[('media', 3), ('work', 1), ('crossroads', 1), ('study', 1), ('ways', 1)]


And now verbs.

In [56]:
# And we'll do the verbs in one fell swoop

verbs = [word for word,pos in tagged_tokens if pos == 'VB' or pos=='VBD' or pos=='VBG' or pos=='VBN' or pos=='VBP' or pos=='VBZ']
verb_frequency = collections.Counter(verbs)
print(verb_frequency.most_common(5))

[("'s", 2), ('bringing', 2), ('bear', 2), ('has', 1), ('do', 1)]


In [57]:
# If we bring all of this together we get a pretty good summary of the sentence

print(adj_frequency.most_common(3))
print(noun_frequency.most_common(3))
print(verb_frequency.most_common(3))

[('digital', 3), ('humanistic', 3), ('traditional', 2)]
[('media', 3), ('work', 1), ('crossroads', 1)]
[("'s", 2), ('bringing', 2), ('bear', 2)]


# 5. Demonstration: Guess the Novel

To illustrate this process on a slightly larger scale, we will do the exactly what we did above, but will do so on two unknown novels. Your challenge: guess the novels from the most frequent words.

We will do this in one chunk of code, so another challenge for you during breaks or the next few weeks is to see how much of the following code you can follow (or, in computer science terms, how much of the code you can parse). If the answer is none, not to worry! Tomorrow we will take a step back and work on the nitty gritty of programming.

In [80]:
# Read the two text files from your hard drive
# Assign first mystery text to variable 'text1' and second to 'text2'

text1 = open('text1.txt').read()
text2 = open('text2.txt').read()

In [81]:
# Tokenize both texts

text1_tokens = nltk.word_tokenize(text1)
text2_tokens = nltk.word_tokenize(text2)

In [82]:
# Set to lower case

text1_tokens_lc = [word.lower() for word in text1_tokens]
text2_tokens_lc = [word.lower() for word in text2_tokens]

In [83]:
# Remove stopwords

text1_tokens_nostops = [word for word in text1_tokens_lc if word not in stopwords.words('english')]
text2_tokens_nostops = [word for word in text2_tokens_lc if word not in stopwords.words('english')]

In [84]:
# Remove punctuation using the list of punctuation from the string pacage

text1_tokens_clean = [word for word in text1_tokens_nostops if word not in string.punctuation]
text2_tokens_clean = [word for word in text2_tokens_nostops if word not in string.punctuation]

In [85]:
# Frequency distribution

text1_word_frequency = collections.Counter(text1_tokens_clean)
text2_word_frequency = collections.Counter(text2_tokens_clean)

In [86]:
# Guess the novel!

text1_word_frequency.most_common(20)

[("'s", 1784),
 ("''", 1581),
 ('``', 1513),
 ('whale', 1083),
 ('one', 900),
 ('like', 575),
 ('upon', 565),
 ('ahab', 495),
 ('man', 490),
 ('ship', 461),
 ('old', 441),
 ('would', 435),
 ('ye', 432),
 ('sea', 381),
 ('though', 363),
 ('yet', 339),
 ('time', 323),
 ('captain', 320),
 ('long', 315),
 ('still', 308)]

In [87]:
# Guess the novel!

text2_word_frequency.most_common(20)

[("''", 1641),
 ('``', 1443),
 ("'s", 692),
 ('elinor', 679),
 ('could', 572),
 ('marianne', 560),
 ('mrs.', 521),
 ('would', 514),
 ('said', 387),
 ('every', 373),
 ('one', 326),
 ('much', 288),
 ('must', 282),
 ('sister', 269),
 ('edward', 258),
 ('dashwood', 249),
 ('mother', 249),
 ('time', 237),
 ('jennings', 234),
 ('know', 229)]

Computational Text Analysis is not simply the processing of texts through computers, but involves reflection on the part of human interpreters. How were you able to tell what each novel was? Do you notice any differences between each novel's list of frequent words?

The patterns that we notice in our computational model often enrich and extend our research questions -- sometimes in surprising ways! What next steps would you take to investigate these novels?

# 6. Concordances and Similar Words using NLTK

Tallying word frequencies gives us a bird's-eye-view of our text but we lose one important aspect: context. As the dictum goes: "You shall know a word by the company it keeps."

Concordances show us every occurrence of a given word in a text, inside a window of context words that appear before and after it. This is helpful for close reading to get at a word's meaning by seeing how it is used. We can also use the logic of shared context in order to identify which words have similar meanings. To illustrate this, we can compare the way the word "monstrous" is used in our two novels.

### Concordance

In [88]:
# Transform our raw token lists in NLTK Text-objects
text1_nltk = nltk.Text(text1_tokens)
text2_nltk = nltk.Text(text2_tokens)

In [89]:
# Really they're no differnt from the raw text, but they have additional useful functions
print(text1_nltk)
print(text2_nltk)

<Text: ETYMOLOGY . ( Supplied by a Late Consumptive...>
<Text: CHAPTER 1 The family of Dashwood had long...>


In [90]:
# Like a concordancer!

text1_nltk.concordance("monstrous")

Displaying 11 of 11 matches:
ong the former , one was of a most monstrous size ... . This came towards us , 
n of the Psalms . `` Touching that monstrous bulk of the whale or ork we have r
ll over with a heathenish array of monstrous clubs and spears . Some were thick
d as you gazed , and wondered what monstrous cannibal and savage could ever hav
that has survived the flood ; most monstrous and most mountainous ! That Himmal
they might scout at Moby Dick as a monstrous fable , or still worse and more de
 Radney . ' '' CHAPTER 55 . Of the Monstrous Pictures of Whales . I shall ere l
ing Scenes . In connexion with the monstrous pictures of whales , I am strongly
ere to enter upon those still more monstrous stories of them which are to be fo
ght have been rummaged out of this monstrous cabinet there is no telling . But 
e of Whale-Bones ; for Whales of a monstrous size are oftentimes cast up dead u


In [91]:
text2_nltk.concordance("monstrous")

Displaying 11 of 11 matches:
 `` Now , Palmer , you shall see a monstrous pretty girl . '' He immediately we
your sister is to marry him . I am monstrous glad of it , for then I shall have
ou may tell your sister . She is a monstrous lucky girl to get him , upon my ho
k how you will like them . Lucy is monstrous pretty , and so good humoured and 
Jennings , `` I am sure I shall be monstrous glad of Miss Marianne 's company ,
usual noisy cheerfulness , `` I am monstrous glad to see you—sorry I could not 
t however , as it turns out , I am monstrous glad there was never any thing in 
so scornfully ! for they say he is monstrous fond of her , as well he may . I s
le that she should be . '' `` I am monstrous glad of it . Good gracious ! I hav
thing of the kind . So then he was monstrous happy , and talked on some time ab
e very genteel people . He makes a monstrous deal of money , and they keep thei


### Contextual Similarity

In [92]:
# Get words that appear in a similar context to "monstrous"

text1_nltk.similar("monstrous")

true contemptible christian abundant few part mean careful puzzled
mystifying passing curious loving doleful gamesome singular
delightfully wise perilous fearless


In [93]:
text2_nltk.similar("monstrous")

very so exceedingly heartily a as good great extremely remarkably
sweet vast amazingly


# Closing Reflection

The methods we have looked at today are the bread-and-butter of NLP. Before moving on, take a moment to reflect on the model of textuality that these rely on. Human language texts are split into tokens. Most often, these are transformed into simple tallies: 'whale' appears 1083 times; "dashwood" appears 249 times. This does not resemble human reading at all! Yet in spite of that, such a list of frequent terms makes a useful summary of the text.

A few questions in closing:
* Can we imagine other ways of representing the text to the computer?
* Why do you think term frequencies are uncannily descriptive?
* What is lost from the text when we rely on frequency information alone?
    * Can context similarity recover some of what was lost?
* What kinds of research questions can be answered using these techniques?
    * What kinds can't?